lib/evaluate/constant.cc

```diff
@@ -66,27 +74,37 @@ static ConstantSubscript SubscriptsToOffset(
   return offset;
 }
 
-static Constant<SubscriptInteger> ShapeAsConstant(
-    const ConstantSubscripts &shape) {
-  using IntType = Scalar<SubscriptInteger>;
-  std::vector<IntType> result;
-  for (auto dim : shape) {
-    result.emplace_back(dim);
-  }
-  return {std::move(result),
-      ConstantSubscripts{static_cast<std::int64_t>(shape.size())}};
+template<typename RESULT, typename ELEMENT>
+Constant<SubscriptInteger> ConstantBase<RESULT, ELEMENT>::SHAPE() const {
+  return AsConstantShape(shape_);
 }
 
 template<typename RESULT, typename ELEMENT>
-Constant<SubscriptInteger> ConstantBase<RESULT, ELEMENT>::SHAPE() const {
-  return ShapeAsConstant(shape_);
+auto ConstantBase<RESULT, ELEMENT>::Reshape(
+    const ConstantSubscripts &dims) const -> std::vector<Element> {
+  std::size_t n{TotalElementCount(dims)};
+  CHECK(!empty() || n == 0);
+  std::vector<Element> elements;
+  auto iter{values().cbegin()};
+  while (n-- > 0) {
+    elements.push_back(*iter);
+    if (++iter == values().cend()) {
+      iter = values().cbegin();
+    }
+  }
+  return elements;
 }
 
 template<typename T>
 auto Constant<T>::At(const ConstantSubscripts &index) const -> Element {
   return Base::values_.at(SubscriptsToOffset(index, Base::shape_));
 }
 
+template<typename T>
+auto Constant<T>::Reshape(ConstantSubscripts &&dims) const -> Constant {
```

Comment:
Here's another case where I don't understand what type "auto" represents by looking at this file.

Reply:
It's the function result type, which is a `Constant` of the same type as the class.  I could have written this using `Constant<T> Constant<T>::Reshape(...)` but I prefer the use of `auto` whenever an outlined member function's type is best defined in terms of type definitions in the class.

Reply:
Thanks for the explanation.  Sounds good.